openedx/features/content_type_gating/partitions.py

```diff
@@ -0,0 +1,192 @@
+
+        # If there is no verified mode, all users are granted FULL_ACCESS
+        if not course_mode.has_verified_mode(modes):
+            return cls.FULL_ACCESS
+
+        course_enrollment = apps.get_model('student.CourseEnrollment')
+
+        # TODO: remove this line
+        # enrollment = course_enrollment.get_enrollment(user, course_key)
+
+        mode_slug, is_active = course_enrollment.enrollment_mode_for_user(user, course_key)
+
+        if mode_slug and is_active:
+            course_mode = course_mode.mode_for_course(
+                course_key,
+                mode_slug,
+                modes=course_mode.modes_for_course(course_key, include_expired=True, only_selectable=False),
+            )
+            if course_mode is None:
+                LOG.error(
+                    "User %s is in an unknown CourseMode '%s' for course %s. Granting full access to content for this user",
+                    user.username,
+                    mode_slug,
+                    course_key,
+                )
+                return cls.FULL_ACCESS
+
+            if mode_slug == COURSE_MODE_AUDIT:
+                # TODO: does the below comment mean there is more work to be done here
+                # Check the user email exceptions here:
+                return cls.LIMITED_ACCESS
+            else:
+                return cls.FULL_ACCESS
+        else:
+            # Unenrolled users don't get gated content
+            return cls.FULL_ACCESS
+
+    @classmethod
+    def create_user_partition(cls, id, name, description, groups=None, parameters=None, active=True):  # pylint: disable=redefined-builtin, invalid-name, unused-argument
+        """
+        Create a custom UserPartition to support dynamic groups.
+
+        A Partition has an id, name, scheme, description, parameters, and a list
+        of groups. The id is intended to be unique within the context where these
+        are used. (e.g., for partitions of users within a course, the ids should
+        be unique per-course). The scheme is used to assign users into groups.
+        The parameters field is used to save extra parameters e.g., location of
+        the course ID for this partition scheme.
+
+        Partitions can be marked as inactive by setting the "active" flag to False.
+        Any group access rule referencing inactive partitions will be ignored
+        when performing access checks.
+        """
+        return ContentTypeGatingPartition(
+            id,
+            unicode(name),
+            unicode(description),
+            [
+                cls.LIMITED_ACCESS,
+                cls.FULL_ACCESS,
+            ],
+            cls,
+            parameters,
+            # N.B. This forces Content Type Gating partitioning to always be active on every course,
+            # no matter how the course xml content is set. We will manage enabling/disabling
+            # as a policy in the LMS.
+            active=True,
```

Comment:
it doesn't seem like this code path is wrapped in the feature flag?  
If that is the case, I think we would want it behind the flag until all this code is fully tested

Reply:
This was added as part of @cpennington 's studio branch (it is already in master).  I'm happy to throw this behind a waffle flag - but I think the three of us should chat about the best approach to doing that.

Reply:
@cpennington thoughts?